lib/router/src/lib.rs

```diff
@@ -208,43 +260,113 @@ where
-                    match svc.poll_ready().map_err(error::Error::from)? {
-                        Async::NotReady => {
-                            unreachable!("load shedding services must always be ready");
+                State::Init(ref mut state) => {
+                    let request = state.request.take().expect("polled after ready");
+                    let target = state.target.take().expect("polled after ready");
+                    let make = state.make.take().expect("polled after ready");
+
+                    let mut cache = match state.locked_cache.poll_lock() {
+                        Async::Ready(lock) => lock,
+                        Async::NotReady => return Ok(Async::NotReady),
+                    };
+
+                    State::Acquired(UnlockedCache {
+                        request: Some(request),
+                        target: Some(target),
+                        make: Some(make),
+                        unlocked_cache: cache,
+                    })
+                }
+                State::Acquired(ref mut state) => {
+                    // FIXME(kleimkuhler): This match arm is a bit of a hack.
+                    // It is a closure that is immediately executed. The
+                    // reason for this is because of the mutable borrow that
+                    // happens when we check if the target already exists in
+                    // the cache. It should no longer be an issue with NLL.
+                    (|| {
+                        let request = state.request.take().expect("polled after ready");
+                        let target = state.target.take().expect("polled after ready");
+
+                        // If the target exists in the cache, get the service
+                        // and call it with `request`.
+                        if let Some(service) = state.unlocked_cache.access(&target) {
+                            return State::Call(Some(request), Some(service.node.value.clone()));
+                        }
+
+                        // Since there wasn't a cached route, ensure that
+                        // there is capacity for a new one
+                        let capacity = state.unlocked_cache.capacity();
+                        match state.unlocked_cache.reserve() {
+                            Async::Ready(reserve) => {
+                                // Make a new service for the route
+                                let make = state.make.take().expect("polled after ready");
+                                let service = LoadShed::new(make.make(&target));
+
+                                // Cache the service and send the request on the route
+                                reserve.store(target, service.clone());
+                                State::Call(Some(request), Some(service))
+                            }
+                            Async::NotReady => {
+                                State::Error(Some(error::NoCapacity(capacity).into()))
```

Comment:
I'm not sure if this should fail when there's no capacity. Instead, I think the future should return `NotReady` here, since additional capacity might be released, and enter a state where we keep trying to reserve a slot in the cache. If we spend too long trying to reserve a router slot, the request will still eventually get killed by the dispatch timeout, so this won't hang forever.

Reply:
As per our offline conversation w/ @olix0r, we've determined that failing fast when there's no capacity is the right thing to do here. Commenting just so that it's recorded for posterity.